src/laser_scan_assembler.cpp

```diff
@@ -117,16 +314,171 @@ class LaserScanAssembler : public BaseAssembler<sensor_msgs::LaserScan>
+    scan_hist_.clear();
+    scan_hist_mutex_.unlock();
+
+    current_req_ = req;
+    ROS_WARN_COND(req.horizontal_resolution == 0, "horizontal_resolution is zero");
+    ROS_WARN_COND(req.vertical_resolution == 0, "vertical_resolution is zero");
+
+    scan_range_buffer_ = cv::Mat::zeros(1000, 1000, CV_16UC1);
+    scan_index_ = 0;
+    height_values_ = cv::Mat::zeros(1000, 0, CV_64FC1);
+
+    subscribe();
+
+    resp.success = true;
+    return true;
+  }
+
+  bool stopCollectionAndAssembleScans2(StopCollectionAndAssembleScans2::Request& req, StopCollectionAndAssembleScans2::Response& resp)
+  {
+    unsubscribe();
+
+    scan_hist_mutex_.lock();
+
+    sensor_msgs::PointCloud cloud;
+    bool success = assembleScanIndices(cloud, 0, scan_hist_.size());
+    resp.success = success;
+    if (success)
+    {
+      ROS_INFO_STREAM_NAMED("stopCollectionAndAssembleScans2", "Cloud has " << cloud.points.size() << " points from " << scan_hist_.size() << " scans");
+      sensor_msgs::PointCloud2 cloud2;
+      sensor_msgs::convertPointCloudToPointCloud2(cloud, cloud2);
+      pointcloud2_pub_.publish(cloud2);
+    }
+    else
+    {
+      ROS_INFO_STREAM_NAMED("stopCollectionAndAssembleScans2", "Could not assemble scans");
+      resp.message = "Could not assemble scans";
+    }
+
+    std::vector<double> sorted_heights;
+
+    // Process range image
+    auto filled_roi = cv::Rect(0, 0, max_scan_length_, scan_index_);
+    ROS_DEBUG_STREAM("Buffer has size " << scan_range_buffer_.size() << "and the filled region of that has size " << filled_roi);
+    cv::Mat cropped_buffer = cv::Mat(scan_range_buffer_, filled_roi);
+    std::multimap<double, std::size_t> reordering;
+    //The cropped buffer needs to be sorted before we can use it: 
+    // the height of the scans might not be monotonically increaseing/decreasing when moving up AND down betwen start and stop
+    cv::Mat sorted_ranges = sortMatRowBy(height_values_, cropped_buffer, sorted_heights, reordering);
+
+    // Generate y_map for use in cv::remap
+    ROS_DEBUG("Create y_map_column");
+    cv::Mat y_map_column = cv::Mat::zeros(current_req_.vertical_resolution, 1, CV_32FC1);  // We'll stretch it later, as all the values in a row are the same for Y
+    auto vertical_step = fabs(current_req_.max_height - current_req_.min_height) / current_req_.vertical_resolution;
+    for (size_t y = 0; y < current_req_.vertical_resolution; y++)
+    {
+      auto height_at_pixel = current_req_.min_height + (y*vertical_step);
+      auto row_in_buffer = findInterpolatedIndex(sorted_heights, height_at_pixel);
+      // current_req_.vertical_resolution - y because we want the pixels with lower height at the bottom of the image of course
+      y_map_column.at<float>(current_req_.vertical_resolution - y, 0) = (float)row_in_buffer;
+      // ROS_INFO_STREAM("y: " << y <<
+      //   ", height_at_pixel: " << height_at_pixel <<
+      //   ", row_in_buffer:" << (float)row_in_buffer);
+    }
```

Comment:
Better to loop over the actual and desired heights once and not use `findInterpolatedIndex` separately each time. Or at least not use `.begin` each time but increase that pointer each loop.